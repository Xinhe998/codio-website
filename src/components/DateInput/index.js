import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import DatePicker from '../DatePicker';
import TextInput from '../TextInput';
import { format } from 'date-fns';
import { IoMdCalendar } from 'react-icons/io'
import useClickOutside from '../../hooks/useClickOutside';

const DateInput = ({
  defaultDate,
  onSelect,
  disabledPastDate,
  title,
  placeholder,
  isOpenDatePicker,
  onFocus,
  onBlur,
  switchHandler,
  required,
}) => {
  const dateInputRef = useRef();
  useClickOutside(isOpenDatePicker, dateInputRef, () => switchHandler(false));
  return (
    <div className="DateInput" ref={dateInputRef}>
      <TextInput
        title={title}
        text={format(defaultDate, 'yyyy/MM/dd')}
        placeholder={placeholder}
        onFocus={onFocus}
        onBlur={onBlur}
        icon={<IoMdCalendar />}
        onChange={() => {}}
        required={required}
      />
      {isOpenDatePicker && (
        <DatePicker
          defaultDate={defaultDate}
          onSelect={onSelect}
          switchHandler={switchHandler}
          disabledPastDate={disabledPastDate}
        />
      )}
    </div>
  );
};

DateInput.propTypes = {
  defaultDate: PropTypes.instanceOf(Date),
  onSelect: PropTypes.func,
  disabledPastDate: PropTypes.bool,
  isOpenDatePicker: PropTypes.bool,
  title: PropTypes.string,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
};

DateInput.defaultProps = {
  defaultDate: '',
  onSelect: null,
  disabledPastDate: true,
  title: '',
  placeholder: '',
  isOpenDatePicker: false,
};
export default DateInput;
