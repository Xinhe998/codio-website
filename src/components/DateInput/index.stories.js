import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import StoryRouter from 'storybook-react-router';
import DateInput from './index';
import storeFs from '../../reducers';

const store = createStore(storeFs);
const stories = storiesOf('DateInput', module);

stories.addDecorator(StoryRouter()).add('default', () => (
  <Provider store={store}>
    <DateInput
      title="日期"
      placeholder="dd/mm/yyyy"
      defaultDate={new Date()}
      // onSelect={setBirth}
      disabledPastDate={false}
      isOpenDatePicker
      // onFocus={() => setIsBirthOnFocus(true)}
      // switchHandler={setIsBirthOnFocus}
      required
      // onBlur={() => setIsBirthOnFocus(false)}
    />
  </Provider>
));
