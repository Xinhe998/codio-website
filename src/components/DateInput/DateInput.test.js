/* eslint-disable func-names */
import 'babel-polyfill';
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { BrowserRouter as Router } from 'react-router-dom';
import {
  shallow, configure, render, mount,
} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import DateInput from './index';

import storeFs from '../../reducers';

const store = createStore(storeFs);

// 以該解析器提供給 Enzyme 做渲染 Component 的設置
configure({ adapter: new Adapter() });

describe('DateInput', function () {
  let testComp;

  beforeEach(() => {
    jest.resetModules();
    this.params = {
      defaultDate: new Date('2019/10/24'),
      onSelect: jest.fn(),
      disabledPastDate: true,
      title: 'DateInput',
      placeholder: 'DateInput_Placeholder',
      isOpenDatePicker: false,
      onFocus: jest.fn(),
      onBlur: jest.fn(),
      switchHandler: jest.fn(),
      required: true,
    };

    this.makeSubject = () => {
      const {
        defaultDate,
        onSelect,
        disabledPastDate,
        title,
        placeholder,
        isOpenDatePicker,
        onFocus,
        onBlur,
        switchHandler,
        required,
      } = this.params;

      testComp = mount(
        <Router>
          <Provider store={store}>
            <DateInput
              defaultDate={defaultDate}
              onSelect={onSelect}
              disabledPastDate={disabledPastDate}
              title={title}
              placeholder={placeholder}
              isOpenDatePicker={isOpenDatePicker}
              onFocus={onFocus}
              onBlur={onBlur}
              switchHandler={switchHandler}
              required={required}
            />
          </Provider>
        </Router>,
      );
      return testComp;
    };
  });
  describe('when it has complete data', () => {
    beforeEach(() => {
      this.subject = this.makeSubject();
    });
    it('should be same as snapshot', async () => {
      const wrapper = render(await testComp);
      expect(wrapper).toMatchSnapshot();
    });
    it('should find className', async () => {
      expect(testComp.find('.DateInput'));
    });
    // 測試按下會觸發 onChange
    it('test onChange', () => {
      testComp.setProps({ isOpenDatePicker: true });
      testComp.update();
      expect(testComp.find('DatePicker'));
    });
  });
});
