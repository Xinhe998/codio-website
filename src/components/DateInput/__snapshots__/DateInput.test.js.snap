// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput when it has complete data should be same as snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "DateInput",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "textinput required",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "textinput__title",
            },
            "children": Array [
              Object {
                "data": "DateInput",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "placeholder": "DateInput_Placeholder",
                "type": "text",
                "value": "2019/10/24",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "textinput__icon",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "fill": "currentColor",
                      "height": "1em",
                      "stroke": "currentColor",
                      "stroke-width": "0",
                      "viewBox": "0 0 512 512",
                      "width": "1em",
                      "xmlns": "http://www.w3.org/2000/svg",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "d": "M368.005 272h-96v96h96v-96zm-32-208v32h-160V64h-48v32h-24.01c-22.002 0-40 17.998-40 40v272c0 22.002 17.998 40 40 40h304.01c22.002 0 40-17.998 40-40V136c0-22.002-17.998-40-40-40h-24V64h-48zm72 344h-304.01V196h304.01v212z",
                        },
                        "children": Array [],
                        "name": "path",
                        "namespace": "http://www.w3.org/2000/svg",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "d": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "d": undefined,
                        },
                      },
                    ],
                    "name": "svg",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "fill": undefined,
                      "height": undefined,
                      "stroke": undefined,
                      "stroke-width": undefined,
                      "viewBox": undefined,
                      "width": undefined,
                      "xmlns": "http://www.w3.org/2000/xmlns/",
                    },
                    "x-attribsPrefix": Object {
                      "fill": undefined,
                      "height": undefined,
                      "stroke": undefined,
                      "stroke-width": undefined,
                      "viewBox": undefined,
                      "width": undefined,
                      "xmlns": "",
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "placeholder": "DateInput_Placeholder",
              "type": "text",
              "value": "2019/10/24",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "textinput__icon",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "fill": "currentColor",
                    "height": "1em",
                    "stroke": "currentColor",
                    "stroke-width": "0",
                    "viewBox": "0 0 512 512",
                    "width": "1em",
                    "xmlns": "http://www.w3.org/2000/svg",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "d": "M368.005 272h-96v96h96v-96zm-32-208v32h-160V64h-48v32h-24.01c-22.002 0-40 17.998-40 40v272c0 22.002 17.998 40 40 40h304.01c22.002 0 40-17.998 40-40V136c0-22.002-17.998-40-40-40h-24V64h-48zm72 344h-304.01V196h304.01v212z",
                      },
                      "children": Array [],
                      "name": "path",
                      "namespace": "http://www.w3.org/2000/svg",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "d": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "d": undefined,
                      },
                    },
                  ],
                  "name": "svg",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "fill": undefined,
                    "height": undefined,
                    "stroke": undefined,
                    "stroke-width": undefined,
                    "viewBox": undefined,
                    "width": undefined,
                    "xmlns": "http://www.w3.org/2000/xmlns/",
                  },
                  "x-attribsPrefix": Object {
                    "fill": undefined,
                    "height": undefined,
                    "stroke": undefined,
                    "stroke-width": undefined,
                    "viewBox": undefined,
                    "width": undefined,
                    "xmlns": "",
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "textinput__title",
              },
              "children": Array [
                Object {
                  "data": "DateInput",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "textinput__icon",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "fill": "currentColor",
                  "height": "1em",
                  "stroke": "currentColor",
                  "stroke-width": "0",
                  "viewBox": "0 0 512 512",
                  "width": "1em",
                  "xmlns": "http://www.w3.org/2000/svg",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "d": "M368.005 272h-96v96h96v-96zm-32-208v32h-160V64h-48v32h-24.01c-22.002 0-40 17.998-40 40v272c0 22.002 17.998 40 40 40h304.01c22.002 0 40-17.998 40-40V136c0-22.002-17.998-40-40-40h-24V64h-48zm72 344h-304.01V196h304.01v212z",
                    },
                    "children": Array [],
                    "name": "path",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "d": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "d": undefined,
                    },
                  },
                ],
                "name": "svg",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "fill": undefined,
                  "height": undefined,
                  "stroke": undefined,
                  "stroke-width": undefined,
                  "viewBox": undefined,
                  "width": undefined,
                  "xmlns": "http://www.w3.org/2000/xmlns/",
                },
                "x-attribsPrefix": Object {
                  "fill": undefined,
                  "height": undefined,
                  "stroke": undefined,
                  "stroke-width": undefined,
                  "viewBox": undefined,
                  "width": undefined,
                  "xmlns": "",
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "placeholder": "DateInput_Placeholder",
                "type": "text",
                "value": "2019/10/24",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "textinput__title",
                },
                "children": Array [
                  Object {
                    "data": "DateInput",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
